### PR TITLE
Add fragmentation strategies for dataset summary generation

### PR DESCRIPTION
**Pull request by surilindur**
This is an initial draft pull request for adding fragmentation strategies that will generate dataset summaries using the VoID vocabulary and the custom membership filter vocabulary for Bloom filters. The PR includes the following changes:

* Added `DatasetSummary` classes for VoID description and Bloom filters.
* Added `FragmentationStrategyDatasetSummary` for VoID description and Bloom filters, that collect descriptions for datasets and serialize them when flushed.
* Added instructions/explanation for the new fragmentation strategies in the readme.

The following other changes are also included, but I can try to move them to another PR if they feel unrelated:

* Updated the GitHub Actions to new versions, because the old ones used a deprecated Node version. Added new Node versions to the CI: linting runs on the latest LTS version, the other ones run on specific versions + the latest one.
* Updated to version 2 of `@rubensworks/eslint-config`, fixed the linter errors associated with the update, and added the relevant rule exceptions in ESLint config. Also replaced exec with test for regular expression testing because it seems to be supported by all the relevant Node versions.
* Added a cast from readable-stream's own Readable to Node's own in the file loader, because the fragmenter uses Node's own streams whereas rdf-parse uses those from readable-stream.
* Replaced `npm` with `yarn` in the scripts of package.json, to allow the scripts to run without having to have the NPM CLI installed. Also replaced `pre-commit` with `husky` to enable the pre-commit hooks to work without the NPM CLI.

Any feedback is welcome! I will keep this as a draft until everything has been fixed.

### CODE REVIEW

**Comment by rubensworks** [comment]
At first glance, this looks pretty good, nice work @surilindur!

However, as you suggested, it would be better to have the repo cleanup stuff (like linting) moved to a separate PR.
It makes it quite difficult to review the changes otherwise.
Could you look into that?
We could either first have the cleanup PR, and then the additions, or the other way around, that's up to you.

In any case, I would wait with merging the summary additions PR until we know for certain that @constraintAutomaton's shapes will fit into this framework as well.

**Comment by surilindur** [comment]
The PR has been split now, but I noticed there is an error in the way the vocabulary is used for the Bloom filters. I will fix it before converting this into not-a-draft PR.

**Comment by surilindur** [comment]
This is now rebased on the work in #20 because it helps make development easier

**Comment by surilindur** [comment]
This still needs some fixes for the use of the membership vocabulary, I just noticed... :sob: I will also keep this as a draft until the shape generation has been tested to work with the same idea.

**Comment by rubensworks** [comment]
Ok, no rush! :-)

**Comment by surilindur** [comment]
This should be ready for review whenever there is time for it. I managed to set up the benchmarks using the branch from the fork, so there is no hurry.

File: README.md [hunk 1]
```diff
@@ -226,6 +226,41 @@ except for predicate1 and predicate2 that will be delegated to the object-based
 }
 ```
 
+#### Dataset Summary Fragmentation Strategies
```
Comment (rubensworks) [inline README.md:229]: Could we have 2 subsections here, one for VoID and one for Bloom?

File: test/unit/strategy/FragmentationStrategyDatasetSummaryVoID-test.ts [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+import { DataFactory } from 'rdf-data-factory';
+import { FragmentationStrategyDatasetSummaryVoID } from '../../../lib/strategy/FragmentationStrategyDatasetSummaryVoID';
+import type { IFragmentationStrategy } from '../../../lib/strategy/IFragmentationStrategy';
+import { DatasetSummaryVoID } from '../../../lib/summary/DatasetSummaryVoID';
+
+const streamifyArray = require('streamify-array');
+
+const DF = new DataFactory();
+
+// Jest.mock('../../../lib/io/ParallelFileWriter');
```
Comment (rubensworks) [inline test/unit/strategy/FragmentationStrategyDatasetSummaryVoID-test.ts:12]: This can go?

**Review by rubensworks — CHANGES_REQUESTED** [review]
Good stuff! No major comments, besides two minor ones.

**Comment by surilindur** [comment]
This is now ready for review again! I have tested that it works with SolidBench.

**Comment by rubensworks** [comment]
Thanks! Released as 2.7.0.

**Pull request merged**